src/Pos/Delegation/Logic.hs

```diff
@@ -171,40 +133,294 @@ initDelegation = do
+    -}
+    transChangeset :: m TransChangeset
+    transChangeset = do
+        let keysMap = HS.fromMap . HM.map (const ())
+        let xPoints :: [PublicKey]
+            xPoints = map GS.dlgEdgeActionIssuer $ HS.toList eActions
+
+        -- Step 1.
+        affected <- mconcat <$> mapM calculateLocalAf xPoints
+        let af = keysMap affected
+
+        -- Step 2.
+        dlgNew <- execStateT (for_ af calculateDlgNew) HM.empty
+        -- Let's check that sizes of af and dlgNew match (they should).
+        -- We'll need it to merge in (3).
+        let dlgKeys = keysMap dlgNew
+        when (dlgKeys /= af) $ throwM $ DBMalformed $
+            sformat ("transChangeset: dlgNew keys "%listJson%" /= af "%listJson)
+                    (HS.toList dlgKeys) (HS.toList af)
+
+        -- Step 3.
+        -- Some unsafe functions (чтобы жизнь медом не казалась)
+        let lookupUnsafe k =
+                fromMaybe (error $ "transChangeset shouldn't happen but happened: " <> pretty k) .
+                HM.lookup k
+            toTheseUnsafe :: PublicKey
+                          -> (Maybe PublicKey, Maybe PublicKey)
+                          -> These PublicKey PublicKey
+            toTheseUnsafe a = \case
+                (Nothing,Nothing) ->
+                    error $ "Tried to convert (N,N) to These with affected user: " <> pretty a
+                (Just x, Nothing) -> This x
+                (Nothing, Just x) -> That x
+                (Just x, Just y)  -> These x y
+        let dlgFin = flip HM.mapWithKey affected $ \a dOld ->
+                         toTheseUnsafe a (dOld, lookupUnsafe a dlgNew)
+
+        pure $ dlgFin
+
+    -- Returns map from affected subtree af in original/G to the
+    -- common delegate of this subtree. Keys = af. All elems are
+    -- similar and equal to dlg(iPk).
+    calculateLocalAf :: PublicKey -> m (HashMap PublicKey (Maybe PublicKey))
+    calculateLocalAf iPk = GS.getDlgTransitive iPk >>= \case
+        Nothing -> pure $ HM.singleton iPk Nothing
+        Just dPk -> do
+            -- All i | i → d in the G. We should leave only those who
+            -- are lower than iPk in the delegation chain.
+            revIssuers <- HS.toList <$> GS.getDlgTransitiveReverse dPk
+            -- For these we'll find everyone who's upper (closer to
+            -- root/delegate) and take a diff.
+            chain <- HM.keys <$> GS.getPskChain (Left iPk)
+            let ret = iPk : (revIssuers \\ chain)
+            let retHm = HM.fromList $ map (,Just dPk) ret
+            -- Just a sanity check. If you're optimizing, delete it.
+            unless (HM.size retHm /= length ret) $ throwM $ DBMalformed $
+                sformat ("calculateLocalAf for iPk "%build%" has duplicates: "%listJson)
+                        iPk ret
+            pure retHm
+
+    eActionsHM :: HashMap PublicKey GS.DlgEdgeAction
+    eActionsHM =
+        HM.fromList $ map (\x -> (GS.dlgEdgeActionIssuer x, x)) $ HS.toList eActions
```

Comment:
Well, it's a `DlgEdgeAction` helper. I think the most appropriate place for it is in `DB` because that's where it is used mostly.